mmgen/models/diffusions/base_diffusion.py

```diff
@@ -0,0 +1,903 @@
+            num_batches (int, optional):
+            label (torch.Tensor, optional):
+            save_intermedia (bool, optional): Whether to save denoising result
+                of intermedia timesteps. If set as True, would return a list
+                contains denoising results of each timestep. Otherwise, only
+                the final denoising result would be returned. Defaults to
+                False.
+            timesteps_noise (torch.Tensor, optional): Noise term used in each
+                denoising timesteps. If passed, the input noise would shape as
+                [num_timesteps, b, c, h, w]. If set as None, noise of each
+                denoising timestep would be randomly sampled. Default as None.
+        Returns:
+            torch.Tensor | list[torch.Tensor]: If ``save_intermedia``, a list
+                contains denoising results of each timestep would be returned.
+                Otherwise, only the final denoising result would be returned.
+        """
+        device = get_module_device(self)
+        x_t = self.get_noise(noise, num_batches=num_batches).to(device)
+        if save_intermedia:
+            # save input
+            intermedia = [x_t.clone()]
+
+        # use timesteps noise if defined
+        if timesteps_noise is not None:
+            timesteps_noise = self.get_noise(
+                timesteps_noise, num_batches=num_batches,
+                timesteps_noise=True).to(device)
+
+        batched_timesteps = torch.arange(self.num_timesteps - 1, -1,
+                                         -1).long().to(device)
+        for t in batched_timesteps:
+            batched_t = t.expand(x_t.shape[0])
+            step_noise = timesteps_noise[t, ...] \
+                if timesteps_noise is not None else None
+
+            x_t = self.denoising_step(
+                model, x_t, batched_t, noise=step_noise, label=label, **kwargs)
+            if save_intermedia:
+                intermedia.append(x_t.clone())
+        if save_intermedia:
+            return intermedia
+        return x_t
+
+    def prepare_diffusion_vars(self):
+        """Prepare for variables used in the diffusion process.
+
+        TODO: we should use cupy or torch for speeding up?
+        """
+        self.betas = self.get_betas()
+        self.alphas = 1.0 - self.betas
+        self.alphas_bar = np.cumproduct(self.alphas, axis=0)
+        self.alphas_bar_prev = np.append(1.0, self.alphas_bar[:-1])
+        self.alphas_bar_next = np.append(self.alphas_bar[1:], 0.0)
+
+        # calculations for diffusion q(x_t | x_0) and others
+        self.sqrt_alphas_bar = np.sqrt(self.alphas_bar)
+        self.sqrt_one_minus_alphas_bar = np.sqrt(1.0 - self.alphas_bar)
+        self.log_one_minus_alphas_bar = np.log(1.0 - self.alphas_bar)
+        self.sqrt_recip_alplas_bar = np.sqrt(1.0 / self.alphas_bar)
+        self.sqrt_recipm1_alphas_bar = np.sqrt(1.0 / self.alphas_bar - 1)
```

Comment:
I'm wondering whether we really need to set so many variables as the attribute of this class. You may keep in mind that an attribute should only be defined or clarified in the `__init__` function. Please carefully think about this point and check the attributes in this function.

Reply:
In this function, we preprocess all variables need for the sampling process.
Otherwise, we need to calculate those variables in each function.